dbt/adapters/bigquery/impl.py

```diff
@@ -805,27 +807,28 @@ def grant_access_to(self, entity, entity_type, role, grant_target_dict):
         """
         Given an entity, grants it access to a permissioned dataset.
         """
+
+        if entity_type == "view":
+            entity = self.get_table_ref_from_relation(entity).to_api_repr()
         conn = self.connections.get_thread_connection()
         client = conn.handle
 
         GrantTarget.validate(grant_target_dict)
         grant_target = GrantTarget.from_dict(grant_target_dict)
-        dataset_ref = self.connections.dataset_ref(grant_target.project, grant_target.dataset)
-        dataset = client.get_dataset(dataset_ref)
+        with _dataset_lock:
+            dataset_ref = self.connections.dataset_ref(grant_target.project, grant_target.dataset)
+            dataset = client.get_dataset(dataset_ref)
 
-        if entity_type == "view":
-            entity = self.get_table_ref_from_relation(entity).to_api_repr()
+            access_entry = AccessEntry(role, entity_type, entity)
+            access_entries = dataset.access_entries
 
-        access_entry = AccessEntry(role, entity_type, entity)
-        access_entries = dataset.access_entries
-
-        if access_entry in access_entries:
-            logger.debug(f"Access entry {access_entry} " f"already exists in dataset")
-            return
+            if access_entry in access_entries:
+                logger.debug(f"Access entry {access_entry} " f"already exists in dataset")
+                return
 
-        access_entries.append(AccessEntry(role, entity_type, entity))
-        dataset.access_entries = access_entries
-        client.update_dataset(dataset, ["access_entries"])
+            access_entries.append(AccessEntry(role, entity_type, entity))
```

Comment:
access_entries.append(access_entry)

Reply:
Unless the access_entry assignment needs to be in the lock, I would move that assignment outside of the lock, and then make the contents of the `with` block a separate function, e.g. `self._update_access_entries(access_entry, grant_target)`.

Reply:
since access entries is tied to the dataset being updated I believe we do actually need to perform this logic as part of a transaction within the lock: aquire lock>get the dataset>get the access entries for the dataset>check if the access entry exists>update the dataset>exit lock